### PR TITLE
[docker] Obey constraints file in ray Docker image

### PR DESCRIPTION
**Pull request by krfricke**
## Why are these changes needed?

The `ray` docker container currently doesn't install with the constraints file specified - meaning that the shipped packages are different from the ones in `ray-ml`.

This PR resolves this issue.

In a follow-up, we should build a script to automatically build environment-aware constraints in a recurring job.

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/requirements_compiled.txt [hunk 1]
```diff
@@ -220,8 +230,8 @@ jsonschema==4.17.3
 jsonschema-spec==0.1.6
 junit-xml==1.9
 jupyter-cache==0.4.3
-jupyter-client==7.3.4
-jupyter-core==5.3.1
+jupyter-client==7.3.4 ; python_version >= "3.8"
+jupyter-core==5.3.1 ; python_version >= "3.8"
```
Comment (aslonnie) [inline python/requirements_compiled.txt:234]: why are these `python_version >= "3.8"` required? what will happen for python 3.7 here?

Comment (aslonnie) [reply python/requirements_compiled.txt:234]: in particular, is it that for python 3.7 these libraries are just not required/installed?

Comment (krfricke) [reply python/requirements_compiled.txt:234]: Some of these dependencies have sub-dependencies that are unavailable for Python 3.7. The `python_version >= "3.8"` effectively removes the constraint for python 3.7.

In this PR, I compiled requirements for python 3.7 separately and added the selector for any constraints that were resolved to a lower version than in 3.8. This seems to resolve fine.

This is not a sustainable practice. While we can try to update single dependencies manually in the short run, we do need a comprehensive compilation mechanism.

Comment (aslonnie) [reply python/requirements_compiled.txt:234]: hmm... maybe should just have a separate file for 3.7 given there are so many differences.

File: ci/build/build-docker-images.py [hunk 2]
```diff
@@ -421,6 +428,19 @@ def build_or_pull_base_images(
         return False
 
 
+def prep_ray_base():
+    root_dir = _get_root_dir()
+    requirements_files = [
+        "python/requirements_compiled.txt",
+        "python/requirements_compiled_py37.txt",
+    ]
```
Comment (aslonnie) [inline ci/build/build-docker-images.py:436]: should this also be conditioned by the python version?

File: ci/ci.sh [hunk 3]
```diff
@@ -91,6 +91,7 @@ NEED_WHEELS="$(_need_wheels)"
 
 compile_pip_dependencies() {
   # Compile boundaries
+  TARGET=${1-requirements_compiled.txt}
```
Comment (aslonnie) [inline ci/ci.sh:94]: nit: double quote:

`TARGET=${1:-requirements_compiled.txt}`

File: python/requirements_compiled.txt [hunk 4]
```diff
@@ -522,31 +541,38 @@ typer==0.9.0
...
+wheel==0.41.0
+widgetsnbextension==3.6.5
 wrapt==1.15.0
 wurlitzer==3.0.3
 xgboost==1.7.6 ; python_version > "3.7"
 xlrd==2.0.1
 xmltodict==0.13.0
-xxhash==3.2.0
-y-py==0.5.9
+xxhash==3.3.0
```
Comment (aslonnie) [inline python/requirements_compiled.txt:567]: these `python_version > "3.7"` can be removed now?

File: python/requirements_compiled_py37.txt [hunk 5]
```diff
@@ -0,0 +1,562 @@
...
+markupsafe==2.1.3
+matplotlib==3.5.3
+matplotlib-inline==0.1.6
+mccabe==0.6.1
+mdit-py-plugins==0.2.8
+memray==1.9.1 ; platform_system != "Windows" and sys_platform != "darwin" and platform_machine != "aarch64"
+minigrid==2.1.1
+mistune==0.8.4
+mlagents-envs==0.28.0
+mlflow==1.30.0 ; python_version <= "3.7"
```
Comment (aslonnie) [inline python/requirements_compiled_py37.txt:266]: these `python_version` constraints can be removed?

**Review by aslonnie — APPROVED** [review]

File: python/requirements_compiled_py37.txt (hunk 5 above)
Comment (krfricke) [reply python/requirements_compiled_py37.txt:266]: This is auto-added by pip-compile

File: python/requirements_compiled.txt (hunk 4 above)
Comment (krfricke) [reply python/requirements_compiled.txt:567]: Auto-added by pip-compile, I believe it will disappear once we get rid of these markers in the input files

**Pull request merged**